pulpcore/pulpcore/tasking/tasks.py

```diff
@@ -99,6 +99,16 @@ def _queue_reserved_task(name, inner_task_id, resources, inner_args, inner_kwarg
     """
     task_status = TaskStatus.objects.get(pk=inner_task_id)
     while True:
+        if name == "pulpcore.app.tasks.orphan.orphan_cleanup":
+            if ReservedResource.objects.exists():
+                # wait until there are no reservations
+                time.sleep(0.25)
+                continue
+            else:
+                celery.tasks[name].apply(args=inner_args, task_id=inner_task_id,
```

Comment:
I had imagined this orphan cleanup code would run synchronously here instead of being routed to a worker. With this design, another piece of work could be routed quickly to another worker and then it's possible the orphan cleanup code isn't running as a singleton. Can this call the tasking code synchronously somehow?

Also assuming the task code run synchronously, for cancellation purposes I believe we would have this take _queue_release_reasource use the user facing task ID so that if a cancel happens it's actually this task (in the resource manager) that will get killed.

Reply:
Oh you are right `apply()` does run locally.

Reply:
Updated `apply_async_with_reservation` to send the inner_task_id as task_id to `_queue_reserved_task`. That fixed task cancel.